keel-core/src/main/kotlin/com/netflix/spinnaker/keel/constraints/ConstraintEvaluator.kt

```diff
@@ -31,6 +33,9 @@ interface ConstraintEvaluator<T : Constraint> {
     artifact: DeliveryArtifact,
     version: String,
     deliveryConfig: DeliveryConfig,
-    targetEnvironment: String
+    targetEnvironment: Environment,
+    priorState: ConstraintState? = null
```

Comment:
Shouldn't this be defined in `StatefulConstraintEvaluator` instead, since it's about storing state?

Reply:
I think the current `canPromote` signature needs a refactor but I'd rather not have a different interface for constraint evaluation depending on type, at least not yet.

Reply:
IIUC, the only use for this is in `EnvironmentPromotionChecker`, where we get _current_ state for the constraint immediately before passing it in to `canPromote`. Maybe we just add `currentState` with a getter in `StatefulConstraintEvaluator` which you can call from the `canPromote` override in sub-classes? That would avoid any references to state in the base interface. Just a thought.

Reply:
Sorry, not a getter, a function like `currentStateFor(<args>)`.

Reply:
@asher I still think `priorState` should be removed from the interface. The `canPromote` method has all the info it needs without it to call `deliveryConfigRepository.getConstraintState(deliveryConfig.name, targetEnvironment.name, version, constraint.type)` to get the current constraint state. I also feel like `EnvironmentPromotionChecker` should not be concerned with the underlying implementation of `ConstraintEvaluators`, including whether they store state or not.

Reply:
Sure, we can just duplicate code in evaluators if you think it's the more elegant route.

Reply:
Like you said earlier, there's really not a super elegant solution, but that would be my vote, yes.